### PR TITLE
Fix calling/invoking through proxies

### PR DESCRIPTION
**Pull request by kasperski95**
- [x] https://github.com/software-mansion/starknet.py/issues/425

### CODE REVIEW

File: tests/data/oz_proxy_compiled_contract.json [hunk 1]
```diff
@@ -0,0 +1 @@
...tarkware.starknet.common.syscalls.delegate_call":{"decorators":[],"pc":3,"type":"function"},"starkware.starknet.common.syscalls.delegate_call.Args":{"full_name":"starkware.starknet.common.syscalls.delegate_call.Args","members":{"calldata":{"cairo_type":"felt*","offset":3},"calldata_size":{"cairo_type":"felt","offset":2},"contract_address":{"cairo_type":"felt","offset":0},"function_selector":{"cairo_type":"felt","offset":1}},"size":4,"type":"struct"},"starkware.starknet.common.syscalls.delegate_call.ImplicitArgs":{"full_name":"starkware.starknet.common.syscalls.delegate_call.ImplicitArgs","members":{"syscall_ptr":{"cairo_type":"felt*","offset":0}},"size":1,"type":"struct"},"starkware.starknet.common.syscalls.delegate_call.Return":{"full_name":"starkware.starknet.common.syscalls.delegate_call.Return","members":{"retdata":{"cairo_type":"felt*","offset":1},"retdata_size":{"cairo_type":"felt","offset":0}},"size":2,"type":"struct"},"starkware.starknet.common.syscalls.delegate_call.SIZEOF_LOCALS":{"type":"const","value":0},"starkware.starknet.common.syscalls.delegate_call.__temp0":{"cairo_type":"felt","full_name":"starkware.starknet.common.syscalls.delegate_call.__temp0","references":[{"a
```
Comment (Arcticae) [inline tests/data/oz_proxy_compiled_contract.json:1]: can't we compile it from source?

File: protostar/starknet_gateway/gateway_facade.py [hunk 2]
```diff
@@ -446,6 +449,33 @@ async def _create_contract_function(
...
+    ):
+        try:
+            try:
+                contract = await Contract.from_address(
+                    address=contract_address,
+                    client=client or self._gateway_client,
+                    proxy_config=True,
+                )
+                return contract
+            except ClientError as err:
```
Comment (Arcticae) [inline protostar/starknet_gateway/gateway_facade.py:465]: variable is not needed

File: protostar/starknet_gateway/gateway_facade.py [hunk 3]
```diff
@@ -446,6 +449,33 @@ async def _create_contract_function(
...
+                        contract_address=contract_address
+                    ) from err
+                raise err
+        except ProxyResolutionError:
+            contract = await Contract.from_address(
+                address=contract_address,
+                client=client or self._gateway_client,
+                proxy_config=False,
+            )
+            return contract
```
Comment (Arcticae) [inline protostar/starknet_gateway/gateway_facade.py:477]: This logic is pretty tangled up - i'd break that down into small methods. Moreover, can we explain a bit why we do things that way? Or make code do that itself instead?

**Review by Arcticae — CHANGES_REQUESTED** [review]

File: tests/data/oz_proxy_compiled_contract.json (hunk 1 above)
Comment (kasperski95) [reply tests/data/oz_proxy_compiled_contract.json:1]: OZ Proxy preset imports various utilities from cairo-contracts. I copied this file from Starknet.py and minimized it.

File: protostar/starknet_gateway/gateway_facade.py (hunk 2 above)
Comment (kasperski95) [reply protostar/starknet_gateway/gateway_facade.py:465]: ?. The variable is used.

File: tests/data/oz_proxy_compiled_contract.json (hunk 1 above)
Comment (mkaput) [reply tests/data/oz_proxy_compiled_contract.json:1]: sounds like a good reason to keep built file.

also this makes tests faster.

File: protostar/starknet_gateway/gateway_facade.py [hunk 4]
```diff
@@ -446,6 +449,33 @@ async def _create_contract_function(
         except KeyError:
             raise UnknownFunctionException(function_name) from KeyError
 
+    async def _get_resolved_contract_function(
```
Comment (mkaput) [inline protostar/starknet_gateway/gateway_facade.py:452]:
```suggestion
    async def _contract_from_possibly_proxy_address(
```

File: protostar/starknet_gateway/gateway_facade.py [hunk 5]
```diff
@@ -446,6 +449,33 @@ async def _create_contract_function(
...
+        self,
+        contract_address: AddressRepresentation,
+        client: Optional[Client] = None,
+    ):
```
Comment (mkaput) [inline protostar/starknet_gateway/gateway_facade.py:456]:
```suggestion
    ) -> Contract:
```

File: protostar/starknet_gateway/gateway_facade.py [hunk 6]
```diff
@@ -446,6 +449,33 @@ async def _create_contract_function(
...
+        try:
+            try:
+                contract = await Contract.from_address(
+                    address=contract_address,
+                    client=client or self._gateway_client,
+                    proxy_config=True,
+                )
+                return contract
```
Comment (mkaput) [inline protostar/starknet_gateway/gateway_facade.py:464]:
```suggestion
                return await Contract.from_address(
                    address=contract_address,
                    client=client or self._gateway_client,
                    proxy_config=True,
                )
```

File: protostar/starknet_gateway/gateway_facade.py (hunk 3 above)
Comment (mkaput) [inline protostar/starknet_gateway/gateway_facade.py:477]:
```suggestion
            return await Contract.from_address(
                address=contract_address,
                client=client or self._gateway_client,
                proxy_config=False,
            )
```

File: protostar/starknet_gateway/gateway_facade.py [hunk 7]
```diff
@@ -446,6 +449,33 @@ async def _create_contract_function(
...
+            except ClientError as err:
+                if "not deployed" in err.message:
+                    raise ContractNotFoundException(
+                        contract_address=contract_address
+                    ) from err
+                raise err
```
Comment (mkaput) [inline protostar/starknet_gateway/gateway_facade.py:470]:
```suggestion
                raise
```

File: protostar/starknet_gateway/gateway_facade.py [hunk 8]
```diff
@@ -446,6 +449,33 @@ async def _create_contract_function(
         except KeyError:
             raise UnknownFunctionException(function_name) from KeyError
 
+    async def _get_resolved_contract_function(
+        self,
+        contract_address: AddressRepresentation,
+        client: Optional[Client] = None,
+    ):
+        try:
+            try:
```
Comment (mkaput) [inline protostar/starknet_gateway/gateway_facade.py:458]: I don't think you need to nest these try blocks

**Review by mkaput — CHANGES_REQUESTED** [review]

File: protostar/starknet_gateway/gateway_facade.py (hunk 2 above)
Comment (Arcticae) [reply protostar/starknet_gateway/gateway_facade.py:465]: it's returned right away

Comment (kasperski95) [reply protostar/starknet_gateway/gateway_facade.py:465]: NVM I thought you meant `err` variable for some reason.

File: protostar/starknet_gateway/gateway_facade.py (hunk 4 above)
Comment (kasperski95) [reply protostar/starknet_gateway/gateway_facade.py:452]: Methods should start with a verb. Renaming to `_get_contract_from_possibly_proxy_address`.

**Review by mkaput — APPROVED** [review]

**Review by Arcticae — APPROVED** [review]

**Pull request merged**